source/agora/network/Client.d

```diff
@@ -559,22 +551,29 @@ public class NetworkClient
 
                 try
                 {
-                    this.log.format(log_level, "Request '{}' to {} failed: {}",
-                        name, this.address[api_idx], ex.message);
+                    this.log.format(log_level, "Request '{}' to {} failed: {} Attempt #{}/{}",
+                        name, conn.address, ex.message, idx, this.max_retries);
                 }
                 catch (Exception ex)
                 {
                     // nothing we can do
                 }
 
+                if (conn.address == Address.init)
+                {
+                    try this.connections = this.connections.remove!(c => c.api == conn.api);
+                    catch (Exception e) log.trace("Failed to remove dead connection");
+                    break;
+                }
```

Comment:
I guess that's why the `dup`. But still, we can't `dup` every time just because there is a chance a connection gets removed. You can just use a `foreach` / `while` loop instead, or swap the inner loops (iterating on connections inside of retry instead of the other way around).

Reply:
It is not only this loop that may modify the connection list tho, another fiber could do that too.

Reply:
This might be solvable with DList and slicing.

Reply:
@Geod24 , forgot to ping you on this. How do you like the last commit?

Reply:
Ok, I'll take a completely different approach on this.

Reply:
Dropped related commit, will have it in its own PR